rollouts/controllers/rollout_controller.go

```diff
@@ -275,13 +275,26 @@ func (r *RolloutReconciler) getPackageDiscoveryClient(rolloutNamespacedName type
 func (r *RolloutReconciler) reconcileRollout(ctx context.Context, rollout *gitopsv1alpha1.Rollout, strategy *gitopsv1alpha1.ProgressiveRolloutStrategy, packageDiscoveryClient *packagediscovery.PackageDiscovery) error {
 	logger := log.FromContext(ctx)
 
+	targetClusters, err := r.store.ListClusters(ctx, rollout.Spec.Targets.Selector)
 	discoveredPackages, err := packageDiscoveryClient.GetPackages(ctx, rollout.Spec.Packages)
 	if err != nil {
 		logger.Error(err, "failed to discover packages")
 		return client.IgnoreNotFound(err)
 	}
 	logger.Info("discovered packages", "count", len(discoveredPackages), "packages", discoveredPackages)
 
+	packageClusterMatcherClient := packageclustermatcher.NewPackageClusterMatcher(targetClusters.Items, discoveredPackages)
+	allClusterPackages, err := packageClusterMatcherClient.GetClusterPackages(rollout.Spec.PackageToTargetMatcher)
+	if err != nil {
+		logger.Error(err, "get cluster packages failed")
+		return client.IgnoreNotFound(err)
+	}
+
+	targets, err := r.computeTargets(ctx, rollout, allClusterPackages)
+	if err != nil {
+		return err
+	}
```

Comment:
I think new flow is more intuitive in my head.

1. compute all the targets
2. Assign targets to waves
3. then, rollout waves